Bluetooth: controller: Fix race between ULL and LLL when initiating connection

wopu-ot opened the pull request:
When scanning, the LLL checks lll->conn to determine whether to send a
connection indication. That field is cleared in ull_master_setup. If
scanning is restarted before ull_master_setup has a chance to execute,
there is a risk that a second connection indication is sent. That in
turn would lead to ull_conn_setup being called a second time, with a
cleared lll->conn, resulting in a null-pointer access.

To avoid this race, a new flag is introduced to keep track of whether a
connection indication has already been sent for the respective
connection.

This change also removes the use of lll_stop, which was used to solve
the same race condition, but did not handle the case that more than one
scanner or advertiser could share the same connection object.

wopu-ot committed:
Bluetooth: controller: Avoid race between ULL and LLL when initiating conn

Use an "initiated" flag in the lll_conn struct to guard the processing
of PDUs related to connection initiation (CONNECT_IND,
AUX_CONNECT_RSP). This avoids races between ULL and LLL when creating
a connection.

Signed-off-by: Wolfgang Puffitsch <wopu@demant.com>

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -925,6 +925,8 @@ static inline int isr_rx_pdu(struct lll_scan *lll, struct pdu_adv *pdu_adv_rx,
 		ret = lll_stop(lll);
```
Could you please analyze the use of `lll_stop`, this interface along the `lll_is_stop` used in `prepare_cb` should handle the race (unless its implementation is flawed) and bypass any subsequent connection indication being transmitted.

wopu-ot replied:
Thanks for the pointer, I think I could make `lll_stop` also work for our LLL. However, I wonder whether the per-LLL stop flag would also work with coded PHY. As far as I understand the code in ll_create_connection(), the normal scanner and the coded PHY scanner point to the same connection object and could race each other?

cvinayak replied:
> the normal scanner and the coded PHY scanner point to the same connection object and could race each other?

Yes, good you point out. This is not addressed today as the initiator for extended advertising is not yet implemented.

Could we use the `lll_stop` in the 1M and CODED_PHY scanning to stop each other's LLL?

wopu-ot replied:
I think it would be possible to have the LLLs stop each other, but I am not sure whether that is the best solution. As far as I can tell, the current uses of `lll_stop` are all related to the start of a connection and could all be replaced with a flag in the `lll_conn` struct. That would mean that we can remove the `is_stop` flag from the general `lll_hdr` and only include a respective flag in `lll_conn`. Sharing of a connection object between LLLs would be taken care of automatically. Should I give that a shot or do you prefer sticking to `lll_stop`?

cvinayak replied:
> Should I give that a shot or do you prefer sticking to lll_stop?

I have a hunch, I will not stick to `lll_stop`. I dont remember now, but I have the feeling use of `lll_stop` for some scenario did not work.

I think this was during testing: https://github.com/zephyrproject-rtos/zephyr/pull/30735
Possibly when central was trying to connect to device as fast as it could. But I am not able remember it to document a symptom or root cause here (now).

Hence, please go ahead with your solutions to the problem at hand here.

My change request here is, could the scanner context hold the flag instead of each of n connection context? (I know, the flag is in an union, but to future-proof and had the flag to the context)

wopu-ot replied:
I have reworked the change now to use a more general "initiated" flag in the connection object and remove the use of lll_stop.

wopu-ot committed:
Bluetooth: controller: Remove use of lll_stop

Remove use of lll_stop and lll_is_stop and rely on "initiated" flag in
lll_conn struct instead.

Signed-off-by: Wolfgang Puffitsch <wopu@demant.com>

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -768,7 +777,7 @@ static inline int isr_rx_pdu(struct lll_scan *lll, struct pdu_adv *pdu_adv_rx,
 #if defined(CONFIG_BT_CENTRAL)
 	/* Initiator */
-	} else if ((lll->conn) &&
+	} else if ((lll->conn) && (!lll->conn->initiated) &&
```
```suggestion
	} else if (lll->conn && !lll->conn->initiated &&
```

wopu-ot replied:
Done.

cvinayak commented on subsys/bluetooth/controller/ll_sw/openisa/lll/lll_scan.c:
```diff
@@ -672,7 +672,7 @@ static inline uint32_t isr_rx_pdu(struct lll_scan *lll, uint8_t devmatch_ok,
 #if defined(CONFIG_BT_CENTRAL)
 	/* Initiator */
-	} else if ((lll->conn) &&
+	} else if ((lll->conn) && (!lll->conn->initiated) &&
```
```suggestion
	} else if (lll->conn && !lll->conn->initiated &&
```

wopu-ot replied:
Done.

cvinayak reviewed (approved):
Minor nitpicking, otherwise LGTM.

Thalley reviewed (approved):
LGTM
